api/src/server/auth.rs

```diff
@@ -96,110 +91,71 @@ pub struct LoginQuery {
-            .map_err(ResponseError::from)?;
-        oidc_client
-            .validate_token(&id_token, None, None)
-            .map_err(ResponseError::from)?;
+        oidc_client.decode_token(&mut id_token)?;
+        oidc_client.validate_token(&id_token, None, None)?;
+
+        // Call to the userinfo endpoint of the provider
+        let userinfo = oidc_client.request_userinfo(&token).await?;
+        Ok((token, userinfo))
     } else {
-        return Ok(None);
+        Err(ResponseError::InvalidCredentials)
     }
-
-    // Call to the userinfo endpoint of the provider
-    let userinfo = oidc_client
-        .request_userinfo(&token)
-        .await
-        .map_err(ResponseError::from)?;
-
-    Ok(Some((token, userinfo)))
 }
 
-/// Makes a new user given a username and updates the provider [UserProvider]
-fn make_user(
+/// Called when a [User] does not yet exist for the [UserProvider].
+/// Will redirect to a page to set their username to make the [User].
+fn init_user(
     user_provider: &UserProvider,
-    username: &str,
+    userinfo: Userinfo,
     conn: &PgConnection,
-) -> User {
-    let user: User = NewUser { username }
-        .insert()
-        .returning(users::all_columns)
-        .get_result(conn)
-        .unwrap();
+) -> Result<(), ResponseError> {
+    // Insert the user into the DB
+    let user: User = NewUser {
+        // TODO: for now this will set username automatically but
+        // we will need to redirect them to set it before making the
+        // user
+        username: &userinfo.email.unwrap()[..20],
+    }
+    .insert()
+    .returning(users::all_columns)
+    .get_result(conn)
+    .map_err(ResponseError::from)?;
+
     // update provider with the user id
-    let update_res = diesel::update(
+    diesel::update(
         user_providers::table
             .filter(user_providers::dsl::id.eq(user_provider.id)),
     )
     .set(user_providers::columns::user_id.eq(user.id))
-    .execute(conn);
-    // TODO: handle update error
-    if update_res.is_err() {
-        panic!("ASSSS");
-    }
-    user
-}
+    .execute(conn)
+    .map_err(ResponseError::from)?;
 
-///Called when a [User] does not yet exist for the [UserProvider]
-///Will redirect to a page to set their username to make the [User]
-fn init_user(
-    user_provider: &UserProvider,
-    userinfo: Userinfo,
-    token: Token,
-    identity: Identity,
-    sessions: web::Data<RwLock<Sessions>>,
-    conn: &PgConnection,
-) -> Result<HttpResponse, actix_web::Error> {
-    // TODO: for now this will set username automatically but
-    // we will need to redirect them to set it before making the
-    // user
-    let username = userinfo
-        .email
-        .clone()
-        .unwrap()
-        .chars()
-        .take(20)
-        .collect::<String>();
-    let user = make_user(&user_provider, &username, conn);
-    login_user(&user_provider, user, token, userinfo, identity, sessions)
+    Ok(())
 }
```

Comment:
So the main reason i had the two funcs was because `init_user` was gonna be called when we wanted to redirect to the username setup and `make_user` would handle the call after the redirect if that makes sense.